### ISSUE TITLE
Preprocessing crashes in multiprocessing code in v1.0.0

### ISSUE THREAD

**Issue by amulder** [OpenNMT/OpenNMT-py#1693]
I experience a crash in the multiprocessing part of the preprocess.py code in **OpenNMT-py v1.0.0** and in the latest version from the master branch. The crash occurs after about 30 seconds.

The crash does not happen in **OpenNMT-py v0.9.0**, when I use the same command and same data.

My preprocess command is:

```
python $OPENNMT_DIR/preprocess.py \
--train_src $DATA_IN_DIR/train_src \
--train_tgt $DATA_IN_DIR/train_tgt \
--valid_src $DATA_IN_DIR/valid_src \
--valid_tgt $DATA_IN_DIR/valid_tgt \
--save_data $DATA_OUT_DIR/preprocessed \
```

The log file output is:

```
[2020-01-02 19:49:33,394 INFO] Extracting features...
[2020-01-02 19:49:33,395 INFO]  * number of source features: 0.
[2020-01-02 19:49:33,395 INFO]  * number of target features: 0.
[2020-01-02 19:49:33,395 INFO] Building `Fields` object...
[2020-01-02 19:49:33,395 INFO] Building & saving training data...
Traceback (most recent call last):
  File "../../../OpenNMT-py-1.0.0/preprocess.py", line 6, in <module>
    main()
  File "/home/ec2-user/test/OpenNMT-py-1.0.0/onmt/bin/preprocess.py", line 293, in main
    preprocess(opt)
  File "/home/ec2-user/test/OpenNMT-py-1.0.0/onmt/bin/preprocess.py", line 273, in preprocess
    'train', fields, src_reader, tgt_reader, align_reader, opt)
  File "/home/ec2-user/test/OpenNMT-py-1.0.0/onmt/bin/preprocess.py", line 200, in build_save_dataset
    for sub_counter in p.imap(func, shard_iter):
  File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 748, in next
    raise value
  File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 431, in _handle_tasks
    put(task)
  File "/usr/local/lib/python3.7/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/usr/local/lib/python3.7/multiprocessing/connection.py", line 393, in _send_bytes
    header = struct.pack("!i", n)
struct.error: 'i' format requires -2147483648 <= number <= 2147483647
```

Any ideas?

**Comment by vince62s** [comment]
what is your pytorch version and do you have a way to try it with python3.6 instead ?

**Comment by francoishernandez** [comment]
@amulder 
Looks quite close to [this](https://stackoverflow.com/questions/47776486/python-struct-error-i-format-requires-2147483648-number-2147483647).
What is your shard size ?
What's strange though is that it appears to have been fixed in python a while back...

**Comment by amulder** [comment]
The crash happens with Python 3.6.10 and Python 3.7.6, and with N=1 and N=4 for num_threads. My pytorch version is 1.3.1 and torchtext is 0.4.0.  I wasn't able to reproduce the crash with the EN-DE data set, nor with the CNN/Daily Mail data set. I am using proprietary data.

@francoishernandez: Thanks! From that link, it looks like this bug fix in cpython has not been applied to 3.6.10 nor has it been applied to 3.7.6. It is fixed in Python 3.8, but there is currently no pytorch release for Python 3.8. 

But, going on @francoishernandez's suggestion, I am able to get past the point where preprocess.py crashed by reducing the shard size from 1000000 (default) to something below roughly 100000. A shard size of 50000 works and the script does not crash. 

If I remember, I can return to this situation when PyTorch is available for Python 3.8 to see if that fixes it. 

Thanks!
